### ISSUE TITLE
Changing a dict in a List(Dict) trait raises a TraitError

### ISSUE THREAD

**Issue by punchagan** [enthought/traits#25]
``` python
from traits.api import *

class A(HasTraits):
    foo = List(Dict)
    bar = Property(List(Str), depends_on='foo[]')

    def _get_bar(self):
        return sorted([sorted(d.keys())[0] for d in self.foo])

a = A()
a.foo.append(dict(x=10))
a.foo[0]['x'] = 20
```

Raises the following TraitError

```
Traceback (most recent call last):
  File "/home/punchagan/tmp/foo.py", line 12, in <module>
    a.foo[0]['x'] = 20
  File "/home/punchagan/work/traits/traits/trait_handlers.py", line 3159, in __setitem__
    raise excp
traits.trait_errors.TraitError: Each value of the 'foo_items' trait of an A instance must be an implementor of, or can be adapted to implement, TraitListEvent or None, but a value of <traits.trait_handlers.TraitDictEvent object at 0x13046d0> <class 'traits.trait_handlers.TraitDictEvent'> was specified.
```

**Comment by corranwebster** [comment]
Related to this (not raising an error, but surprising):

from traits.api import HasTraits, Dict, Str, List, Int

    class TestList(HasTraits):
        x = List(List(Int))

    class TestDict(HasTraits):
        x = Dict(Str, List)
 
    class TestDict2(HasTraits):
        x = Dict(Str, Dict(Str, Int))

def listener(obj, name, old, new):
    print obj, name, old, new.added

    t1 = TestList(x=[[1, 2, 3]])
    t1.on_trait_change(listener, name='x_items')
    t1.x[0].append(4)

    t2 = TestDict(x={'a': [1, 2, 3]})
    t2.on_trait_change(listener, name='x_items')
    t2.x['a'].append(4)

    t3 = TestDict2(x={'a': {'b': 1}})
    t3.on_trait_change(listener, name='x_items')
    t3.x['a']['c'] = 2

The root issue is with recursive calls to `List.validate` or `Dict.validate` when building the `TraitListObject` or `TraitDictObject` instances.  This is non-trivial to fix: it's not even clear what the "right" behaviour should be.

**Comment by jonathanrocher** [comment]
I ran into the same issue (in the `Dict(Str, List)` case). To work around it, @Bouke , I simplified my declaration to `Dict()`. @rkern also suggested offline to use Dict(Str, Instance(list)) if you need the typechecking.

**Comment by mdickinson** [comment]
This is essentially the same issue as #281. We should close one or the other.

**Comment by mdickinson** [comment]
Closed #281 as a duplicate of this issue; there's some useful discussion in comments on that issue.
